provider/oracle/bootstrap.go

```diff
@@ -0,0 +1,172 @@
+	"github.com/juju/juju/instance"
+)
+
+// launchBootstrapController creates a new vm inside
+// the oracle infrastracuture and parses the response into a instance.Instance
+func launchBootstrapConstroller(c *oci.Client, params []oci.InstanceParams) (instance.Instance, error) {
+	if c == nil {
+		return nil, errors.NotFoundf("client")
+	}
+
+	logger.Infof("Launching tbe bootstrap creation method")
+	resp, err := c.CreateInstance(params)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	fmt.Printf("%v\n", resp)
+	os.Exit(1)
+	instance, err := newInstance(resp)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return instance, nil
+}
+
+// uploadSSHControllerKeys checks if the oracle api has the present key if not
+// then it will upload it, if there's arealdy the key we should check if the key
+// there maches the key we want to add, if not then we should modify the content
+// in order to be sure we are using the correct key
+func uploadSSHControllerKeys(c *oci.Client, key string) (string, error) {
+	if c == nil {
+		return "", errors.Errorf("cannot use nil error client to upload ssh keys")
+	}
+
+	if len(key) == 0 {
+		return "", errors.Errorf("cannot use empty key in order to upload it")
+	}
+
+	resp, err := c.SSHKeyDetails("juju-client-key")
+	if err != nil {
+		if strings.Contains(err.Error(), "does not exist") {
+			// the keys does not exist, we should upload them
+			resp, err = c.AddSHHKey("juju-client-key", key, true)
+			if err != nil {
+				return "", errors.Errorf("cannot upload the juju client keys")
+			}
+		} else {
+			return "", errors.Errorf("cannot get details of the juju-client-key")
+		}
+	}
+
+	keyrsponse := key[:strings.Index(key, "juju-client-key")-1]
+	if !resp.Enabled || resp.Key != keyrsponse {
+		if !resp.Enabled {
+			logger.Debugf("The Enabled flag is set to false on the key, setting it to true")
+		}
+		if resp.Key != keyrsponse {
+			logger.Debugf("The key provided with the oracle one does not match, update the key content")
+		}
+		resp, err = c.UpdateSSHKey("juju-client-key", key, true)
+		if err != nil {
+			return "", errors.Errorf("cannot enable of update the juju-client-key content")
+		}
+		if !resp.Enabled {
+			return "", errors.Errorf("juju client key is not enabled after previously tried to enable")
+		}
+	}
+
+	return resp.Name, nil
+}
+
+// isComplaint takes a ImageList object and finds
+// if it contains the given shapName, imageVersion and imageArch
+// if not it will return false
+func isComplaint(
+	list response.ImageList,
+	shapeName, imageVersion, imageArch string,
+) bool {
+
+	i := 0
+	for _, val := range list.Entries {
+		if strings.Contains(
+			val.Attributes.SupportedShapes,
+			shapeName,
+		) {
+			i++
+			break
+		}
+	}
+	if strings.Contains(list.Name, imageVersion) {
+		i++
+	}
+	if strings.Contains(list.Name, imageArch) {
+		i++
+	}
+	if i == 3 {
+		return true
+	}
+
+	return false
+}
+
+// checkImageList checks if there is a image specified in the
```

Comment:
> Ok, so let's clarify some things first. Let me tell what I already understand. Bare with me a moment. Sorry for these kind of mistakes or being a total newbie.

No problem at all, it's not a simple thing to write a provider :)   Also my understanding of Oracle Cloud is not clear, so take what I say with a grain of salt.

> I must add that some public providers has these endpoints defined somewhere and their are fixed(for every region), but in the oracle provider, for every account there is a different kind of endpoint, so the endpoint somehow is nondeterministic(like other providers)

If this alone were the issue, I would think we could come up with image metadata that encoded the common part of the endpoint (i.e. strip off the account-specific prefix). However...

> and on top of that, oracle does not have ubuntu images on default installed and you need to manually surf the marketplace and add those in your account or you have the option to upload them into your account.. Oracle cloud acts more like a private cloud rather than a public one from my point of view.

OK, in that case I'm not sure that we want to try to use the public image metadata at all. If you drop the "Region() (simplestreams.CloudSpec, error)" method from the Environ implementation, Juju will not attempt to use image metadata from cloud-images.ubuntu.com. Instead, it will be up to the provider to identify the appropriate images.

So I suppose what you should do is:
 - drop the Region method, as mentioned above
 - change the "checkImageList" function so that it returns all available images that Juju could potentially use
 - use the environs/instances.FindInstanceSpec function as described elsewhere

That's essentially what the Azure provider does today. The problem with this approach is that the mapping between image IDs and arch/series gets hard-coded into the provider, which means if something new comes along, or if the ID format changes, the provider is missing out on series or is broken. For that reason, we are working on moving Azure over to using image metadata. It sounds like that's not an option for Oracle, though.